### ISSUE TITLE
weekly useful materials - 04/13 -

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#96]

**Comment by GENZITSU** [comment]
## [Transformerモデルの高速化](https://tech.retrieva.jp/entry/20220406)

Tranformerの推論速度がOpenVINO(CPU)とTensorRT(GPU)を用いてどれだけ高速化を検証した記事。

いずれの形式でもonnxにconvertしたモデルを変換することで推論が可能。

OpenVINO形式はonnxのcpu推論より遅い結果になったが、TensorRTはonnxのgpu推論よりも高速化したとのこと。

![スクリーンショット 2022-04-09 19 13 06](https://user-images.githubusercontent.com/19774756/162567551-2d48835a-2264-483b-a204-1ec00c73e1d5.png)

![スクリーンショット 2022-04-09 19 13 17](https://user-images.githubusercontent.com/19774756/162567558-d6e2aec2-64bb-42dc-8f16-8bdce2f2c97e.png)

### コメント
メモ

### 出典
[Transformerモデルの高速化](https://tech.retrieva.jp/entry/20220406)

**Comment by GENZITSU** [comment]
## 機械学習モデルを軽量化せよ！Tensorflow Liteのモデル最適化について

Tensorflow Liteを用いて実施できる量子化や枝刈りを行うことで、精度とモデルサイズがどのように変化するかを検証した記事

float32をint8にしても意外に精度が下がらないという結果を得ている。

![スクリーンショット 2022-04-09 19 27 16](https://user-images.githubusercontent.com/19774756/162568003-f2432d9d-6ee9-4e86-bcff-f86f3d7f9a32.png)

### コメント
メモ

### 出典
[機械学習モデルを軽量化せよ！Tensorflow Liteのモデル最適化について](https://recruit.gmo.jp/engineer/jisedai/blog/tensorflow_lite_optimization/)

**Comment by GENZITSU** [comment]
## [informatix-inc](https://github.com/informatix-inc) / [bert](https://github.com/informatix-inc/bert)

informatix社が日本語コーパスで事前学習を行なったrobertaがapach-2.0ライセンスにて公開。

> his repository provides snippets to use RoBERTa pre-trained on Japanese corpus. Our dataset consists of Japanese Wikipedia and web-scrolled articles, 25GB in total. 

> We trained our model in the same way as RoBERTa. We optimized our model using the masked language modeling (MLM) objective. The accuracy of the MLM is 72.0%.

使い方に若干の癖がありそう

```
# from https://github.com/informatix-inc/bert

    # Paths to each file
    bpe_file = <Path to the file (bpe.txt) which defined word pieces>
    count_file = <Path to the file (bpe_dict.csv) which defines ids for word pieces>
    roberta_config_path = <Path to the file (roberta_config.json) which defines configurations of RobertaModel>
    juman_config_path = <Path to config file for juman>
    
    roberta_weight_path = <Path to the weight file (roberta.pth) of RobertaModel>
    linear_weight_path = <Path to the weight file (linear_word.pth) of final linear layer for MLM>
    
    # load tokenizer
    processor = TextProcessor(bpe_file=bpe_file, count_file=count_file)

    # load pretrained roberta model
    with open(roberta_config_path, "r") as f:
        config_dict = json.load(f)
    config_bert = RobertaConfig().from_dict(config_dict)
    roberta = RobertaModel(config=config_roberta)
    roberta.load_state_dict(torch.load(roberta_weight_path, map_location=device))

    # load pretained decoder
    ifxroberta = IFXRoberta(roberta)
    ifxroberta.linear_word.load_state_dict(torch.load(linear_weight_path, map_location=device))
```

### コメント
メモ

### 出典
[informatix-inc](https://github.com/informatix-inc) / [bert](https://github.com/informatix-inc/bert)

**Comment by GENZITSU** [comment]
## Announcing AWS Lambda Function URLs: Built-in HTTPS Endpoints for Single-Function Microservices

AWS LambdaがAPI GatewayなしでそのままAPIとして叩けるようになったとのこと

> Today, I’m happy to announce the general availability of Lambda Function URLs, a new feature that lets you add HTTPS endpoints to any Lambda function and optionally configure [Cross-Origin Resource Sharing (CORS) headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS).

### コメント
メモ

### 出典
[Announcing AWS Lambda Function URLs: Built-in HTTPS Endpoints for Single-Function Microservices](https://aws.amazon.com/jp/blogs/aws/announcing-aws-lambda-function-urls-built-in-https-endpoints-for-single-function-microservices/)

**Comment by GENZITSU** [comment]
## バンダイナムコグループで対照学習による埋め込み表現を用いてレコメンド技術検証をした話

アニメのシリーズ　関係を用いた対象学習により、BERTをfinetuningすることで、レコメンド性能が高められないかを検証した記事。

![スクリーンショット 2022-04-11 21 39 55](https://user-images.githubusercontent.com/19774756/162741113-65f3a538-c8f4-4499-81d7-82599044b2ae.png)

上記のようなfine-tuningをすることで、fine-tuninigなしと比べて、より良いレコメンドが行えることを示唆している。

![スクリーンショット 2022-04-11 21 40 13](https://user-images.githubusercontent.com/19774756/162741297-db2507d8-adce-4567-9f8d-e267e62c275e.png)

![スクリーンショット 2022-04-11 21 40 20](https://user-images.githubusercontent.com/19774756/162741315-685162e3-254f-4ced-828b-64dc108ecdc7.png)

![スクリーンショット 2022-04-11 21 41 28](https://user-images.githubusercontent.com/19774756/162741375-68fc8f2a-24a4-4898-8939-c325fa1e48e3.png)

### コメント
アニメだとどうしても画風や制作会社というものが影響することもありそうなので、サムネイル画像とかも一緒にembeddingするとより良くなるかも？

### 出典
[バンダイナムコグループで対照学習による埋め込み表現を用いてレコメンド技術検証をした話](https://www.wantedly.com/companies/company_9704487/post_articles/390570)

**Comment by GENZITSU** [comment]
## [LAION-5B: A NEW ERA OF OPEN LARGE-SCALE MULTI-MODAL DATASETS](https://laion.ai/laion-5b-a-new-era-of-open-large-scale-multi-modal-datasets/)

[LAION-400M](https://github.com/GENZITSU/UsefulMaterials/issues/68#issuecomment-918790377)が14倍の5 billion枚のデータセットになって進化した。

![スクリーンショット 2022-04-11 22 01 23](https://user-images.githubusercontent.com/19774756/162744822-5c991a6d-3adc-4357-9adc-32246bd4f322.png)


商用利用可能な[Creative Common CC-BY 4.0](https://creativecommons.org/licenses/by/4.0/)で公開されているものの、商用利用は勧めていない。
> we however do not recommend using it for creating ready-to-go industrial products, as the basic research about general properties and safety of such large-scale models, which we would like to encourage with this release, is still in progress.

LAION-400Mでは512*512以下の小さめの画像ばかりだったのだが、今回はそれを超えるFullHDサイズの画像が増えている。

![スクリーンショット 2022-04-11 21 55 09](https://user-images.githubusercontent.com/19774756/162743775-8adfe138-5790-4d7a-94f7-595afb2118ed.png)

common crawlで収集した画像に対して以下のようなフィルタリングをすることで、データセットを構築している。
> After downloading the WAT files from Common Crawl, we apply the following filtering conditions:
> - All samples with less than 5 characters alt-text length or less than 5 KB image size are dropped.
> - All images with the too big resolution, potentially DOS bombs, were dropped before attempting to process them.
> - Duplicate removal is performed with a bloom filter based on URL. Future runs would include more variate deduplication rules, such as URL + language for the multilanguage dataset.
> - We use CLIP respectively MCLIP to compute embeddings of the image and alt-text. Then we compute the cosine similarity of both embeddings and drop all samples with cosine similarity below 0.28 for the English language ( with CLIP B/32) and 0.26 for the multilingual dataset (MCLIP). These thresholds were selected based on human inspection of the test results.
> - We use the CLIP embeddings of images and texts to filter out to the possible extent the illegal content.

### コメント
まさか一年も経たぬ間に、データセットが14梅になるとは思わなかった...

デモのサイトではテキストで画像を検索できるようになったので、各プロジェクトで必要な画像をピンポイントで探すこともできるかもしれない。

### 出典
- [LAION-5B: A NEW ERA OF OPEN LARGE-SCALE MULTI-MODAL DATASETS](https://laion.ai/laion-5b-a-new-era-of-open-large-scale-multi-modal-datasets/)
- [デモサイト](https://rom1504.github.io/clip-retrieval/?back=https%3A%2F%2Fknn5.laion.ai&index=laion5B&useMclip=false)

**Comment by GENZITSU** [comment]
## [自作MLプロジェクトでMLOps界隈の技術を試してみる その1](https://recruit.gmo.jp/engineer/jisedai/blog/my-mlops-1/)

GCPを使い倒して、モデル学習用のデータ作成ETLからサービング、モニタリングまでを行っている。

![スクリーンショット 2022-04-11 22 59 46](https://user-images.githubusercontent.com/19774756/162755905-21177ce5-a8bf-4edf-ac49-b75fabc395a8.png)

### コメント
参考になりそう

### 出典
[自作MLプロジェクトでMLOps界隈の技術を試してみる その1](https://recruit.gmo.jp/engineer/jisedai/blog/my-mlops-1/)

**Comment by GENZITSU** [comment]
## 最適輸送入門

最適輸送距離を導入するメリットや、実際の解き方などがまとめられている資料

![スクリーンショット 2022-04-11 23 23 34](https://user-images.githubusercontent.com/19774756/162760687-50e8b04f-3c1f-4902-95ce-2975e09b2957.png)

![スクリーンショット 2022-04-11 23 25 07](https://user-images.githubusercontent.com/19774756/162761304-f3afd81c-bbcd-4915-8165-9f2bdc5c8e8d.png)

![スクリーンショット 2022-04-11 23 25 15](https://user-images.githubusercontent.com/19774756/162761323-c4d44368-417c-4702-98ee-8afee6bdc579.png)

![スクリーンショット 2022-04-11 23 25 26](https://user-images.githubusercontent.com/19774756/162761377-2b27e1d8-94b2-419f-bca1-bb1c4eee9cf5.png)

![スクリーンショット 2022-04-11 23 25 43](https://user-images.githubusercontent.com/19774756/162761414-42f2d6e8-d6a9-4c4a-b383-f0e54846f04b.png)

![スクリーンショット 2022-04-11 23 25 55](https://user-images.githubusercontent.com/19774756/162761427-904e1122-768e-4733-9308-a2f66bd37541.png)

### コメント
これまで謎に包まれていたシンクホーンアルゴリズムがなんとなくわかったので、よかった。

### 出典
[最適輸送入門](https://www.slideshare.net/joisino/ss-251328369)

**Comment by GENZITSU** [comment]
## PytorchのDataLoaderの高速化のコツについてすこし解説

pytorchのDataLoaderで何が行われているかの解説から始まり、高速化のためのtipsが紹介されている記事。

高速化については[この記事](https://qiita.com/sugulu_Ogawa_ISID/items/62f5f7adee083d96a587)以上のことは出てきていないが、目を見張るのは仕組み部分。

![スクリーンショット 2022-04-12 0 00 06](https://user-images.githubusercontent.com/19774756/162768317-707e8505-7a00-40bb-9f4e-5959e138f931.png)

> DataLoaderは、iterでアクセスされたタイミング(＿MultiPorcessingDataLoaderIterが生成されるタイミング）で、ワーカーやデータ通信のためのキューを用意します。構成は下記の図のようになっていて、ワーカーごとに入力キュー、ロード済みのデータを読む共通のキュー、ワーカーで構成されています。ワーカー達が準備されたタイミングで、ワーカーのキューに予めnum_workers * prefetch_factor個の処理を突っ込むという処理になっています。
ワーカーへの割当はラウンドロビンとなっています。

![スクリーンショット 2022-04-12 0 00 22](https://user-images.githubusercontent.com/19774756/162768351-c8349599-1c69-4753-9d17-756f66d1a00c.png)

> ＿MultiPorcessingDataLoaderIterにnextメソッドでアクセスしてデータを取り出す際は、インデックスのワーカーのデータが到着するまで待ち、そのデータが得られたらデータを返すという処理になっています。samplerが生成したインデックスの順番を保証するようになっていて、次のインデックスよりも後のインデックスが先に得られた場合は、順番が先のインデックスが得られるまで待ちます。また、データを待つ時間にその際にタイムアウトを設定することもできますが、使うことはあまり無いと思います。

samplerが生成したindex順になるように待つというところが何かしらのボトルネックになるかもしれない。

### コメント
メモ

### 出典
[PytorchのDataLoaderの高速化のコツについてすこし解説](https://zenn.dev/xiongjie/articles/0ae1378feb4204)
